### PR TITLE
feat(open-loop): expand scenario-based open-loop metrics

### PR DESCRIPTION
**Pull request by TatsukiNishimura**
## Summary

This PR expands scenario-based open-loop evaluation with additional metrics for assessing scenario-specific driving behavior and safety.

## Evaluation metrics

The following table describes the intended evaluation scope for each label and
the metrics currently implemented in this PR.

| Eval Label | Assumed scene | Main evaluation metrics | Abnormal cases | Normal cases | Example |
|---|---|---|---|---|---|
| `lane_follow` | Normal lane following, including curb or guardrail proximity, curve overshoot, corner cutting, lane departure, and excessive lateral offset. | `average_lateral_error_m`, `final_lateral_error_m` against the route centerline over an 8.0-second horizon. Maximum lateral deviation is not currently output. | Lane departure, curb or guardrail approach, curve overshoot, corner cutting, or excessive lateral offset. | Normal lane following. |fa0a7e81-430d-4af2-93e1-7d323386a701_0_00000000_00000249 |
| `simple_turn` | Unconstrained normal left- or right-turn scenes without pedestrians, oncoming vehicles, or traffic-light constraints. | `average_lateral_error_m`, `final_lateral_error_m` against the GT ego trajectory over an 8.0-second horizon. This is a lateral-deviation metric, not trajectory ADE/FDE. | Generally out of scope; complex constrained turns should use another label. | Normal left or right turn that follows the GT trajectory. |1184d188-a69b-4bbe-9c53-cdea480a2ca8_0_00000000_00000250|
| `traffic_light_stop` | The ego vehicle should stop and remain stopped because of a red traffic light. The signal is considered red at the start of the override interval.| `success_rate_percent`; pass if `overshoot_m <= 0.5m`. Details: `overshoot_m`, `gt_stop_position_s_m`, `predicted_stop_position_s_m`, `gt_sustained_stop`, `predicted_sustained_stop`. | Moving through the red light, failing to stop, or stopping beyond the GT stopping position. | Stopping and remaining stopped at an appropriate position. | c98f05b9-3dcb-47b8-9e0b-e56adbf58de3_0_00000000_00000250|
| `traffic_light_go` | A traffic-light-related stop constraint is released, and the ego vehicle should start moving after the signal turns green. | `success_rate_percent`;pass if displacement from `ego_current_state` reaches at least 2.0m within 3.0 seconds. Details: `max_displacement_m`, `departed`. Displacement is Euclidean, not limited to the ego forward axis. | Failing to start, remaining stationary, or becoming stuck after release. | Starting normally after the signal changes. |fe0a18b7-e561-4aa2-aeb9-7feaa99475a2_0_00000000_00000249 |
| `pedestrian_yield` | The ego vehicle should stop or yield to a pedestrian or cyclist during a turn or at an intersection without signal control. | `success_rate_percent`; pass if maximum forward progress within 3.0 seconds is at most 0.5m. Details: `max_forward_progress_m`, `yielded`. | Entering the conflict area or starting despite the pedestrian or cyclist. | Waiting appropriately while the pedestrian or cyclist has priority. |ca3f48bb-2526-40bf-998d-e23670d967a0_0_00000000_00000249 |
| `vehicle_yield` | The ego vehicle should yield to another vehicle in an unprotected right turn, merge, or crossing-traffic scene. | Same `yield_progress` implementation as `pedestrian_yield`: `success_rate_percent` based on maximum forward progress of at most 0.5m within 3.0 seconds. Details: `max_forward_progress_m`, `yielded`. | Enteringdespite an oncoming or crossing vehicle, or failing to wait for a safe gap. | Waiting until a safe gap is available. | 3aad7c2c-934c-48ce-84db-f188ab16558f_0_00000000_00000249|
| `obstacle_stop` | The ego vehicle should stop before a static obstacle or stopped vehicle. Close-following and excessive approach scenes are included; safe avoidance andpassing are out of scope. | Same stop-overshoot implementation as `traffic_light_stop`: `success_rate_percent` with a 0.5m overshoot tolerance, plus stop-position and sustained-stop details. | Approaching too closely, overshooting the appropriate stopping position, or failing to stop. | Stopping at an appropriate distance before the obstacle. | c1c33ba9-118d-4dba-a748-69de1518f564_0_00000000_00000249|
| `object_avoidance` | The ego vehicle should safely pass around a parked vehicle, construction object, sign, plant, or other object. Cases with following or oncoming traffic in the avoidance path are included. | `success_rate_percent`; fail if the minimum signed OBB clearance across all valid neighbors is `<= 0.0m`. Details: `min_clearance_m`, `collision`, `status_code`. | Collision, contact with the object, or failure to generate a safe avoidance trajectory. | Safely avoiding the object and continuing through the scene. | 22d04af1-35e0-4a99-9e32-cf30e0dc26e7_0_00000000_00000249|
| `temporal_stop` | The ego vehicle must come to a complete stop at a stop sign or stop line. | Same progress-based implementation as `pedestrian_yield`: `success_rate_percent`; pass if maximum forward progress within 3.0 seconds is at most 0.5m. This currently does not implement a speed-based `complete_stop_success` check before the stop line. | Ignoring the stop requirement or advancing more than 0.5m. | Remaining within the permitted forward-progress tolerance. | |
| `departure` | The ego vehicle should restart after a non-traffic-light stop constraint is released, such as after a pedestrian or oncoming vehicle passes or when departing a bus stop. | `success_rate_percent`; pass if displacement from `ego_current_state` reaches at least 2.0m within 3.0 seconds. Details: `max_displacement_m`, `departed`. As with `traffic_light_go`, this is Euclidean displacement rather than forward-only distance. | Failing to restart after release or becoming stuck. | Restarting normally after the constraint is removed. | 273542e5-0eea-41e4-9846-b02811861d02_0_00000000_00000249|

### Metric output note

The implementation currently reports pass/fail results using
`success_rate_percent`. Some prior specifications used the name
`failure_rate_percent`; that terminology is not used by the current branch.

For `lane_follow` and `simple_turn`, the current implementation outputs average
and final lateral deviation only. It does not currently output maximum lateral
deviation or trajectory ADE/FDE.

## Main changes

- Added eight scenario-based open-loop metrics:
  - `traffic_light_go`
  - `simple_turn`
  - `object_avoidance`
  - `pedestrian_yield`
  - `vehicle_yield`
  - `temporal_stop`
  - `obstacle_stop`
  - `traffic_light_stop`
- Added shared utilities for:
  - Configurable evaluation horizons
  - Lateral and longitudinal trajectory deviation
- Added per-sample diagnostic details to the evaluation output.
- Changed scenario-based evaluation outputs from failure-rate terminology to success-rate terminology.
- Added scenario-specific configuration parameters to `ScenarioOpenLoopConfig`.
- Updated the scenario-based open-loop documentation.
- Changed visualization filenames to use the source NPZ filename, making generated images easier to identify.
- Fixed resolution of relative `--output_dir` paths in `valid_run.py`.

## Input data requirements

Depending on the metric, the following fields may be required:

- `ego_agent_future`
- `route_lanes` or `lanes`
- `neighbor_agents_future`
- `neighbor_agents_past`
- `ego_shape`

## Compatibility note

The `departure` metric output has changed from:

```text
failure_rate_percent
```

to:

```text
success_rate_percent
```

Consumers of the previous output key may need to be updated.

## Tests

```text
17 passed
```

The following areas are covered:

- Scenario-based open-loop configuration and metric registration
- Departure success rate
- Ground-truth trajectory lateral deviation
- Object avoidance
- Stop-position overshoot
- Yield progress

## others
Scorers now receive their input via a new planner_metrics.scene_data.extract_metric_scene_data, which pulls only the fields each metric needs (ego_current_state, ego_agent_future, route_lanes/lanes, neighbor_agents_future, neighbor_agents_past, ego_shape) from a plain mapping, instead of the model's normalized batch. Any source providing these keys can now be scored, not just NPZ. Behavior is unchanged; covered by planner_metrics/test_scene_data.py.

Note: run_scenario_based_open_loop_validation still loads scenario lists as NPZ (dataset loader is out of scope for this PR).

### CODE REVIEW

**Review by xtk8532704 — APPROVED** [review]
LGTM.

File: diffusion_planner/diffusion_planner/scenario_based_open_loop/open_loop.py [hunk 1]
```diff
@@ -214,7 +231,7 @@ def run_scenario_based_open_loop_validation(
                     encoding="utf-8",
                 )
 
-            summaries[metric_name] = {"sample_count": float(count)}
```
Comment (xtk8532704) [inline diffusion_planner/diffusion_planner/scenario_based_open_loop/open_loop.py:217]: **[P1] Removing `sample_count` breaks training-time validation.** `scenario_based_open_loop_validate()` still evaluates `values["sample_count"]` when printing the summary. Every non-empty training-time validation run will therefore raise `KeyError` after the W&B log call. Please either preserve `sample_count` here or update all consumers together.

File: planner_metrics/object_avoidance.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        past = past.to(device=ego_trajs.device)
+        nonzero_future = future[..., :2].abs().sum(dim=-1).gt(0).any(dim=-1)
+        if not nonzero_future.any():
+            raise ValueError(f"object_avoidance found no valid neighbor for sample {index}")
+
+        future = future[nonzero_future]
+        neighbor_shapes = past[nonzero_future, -1, 6:8].to(
+            device=ego_trajs.device, dtype=ego_trajs.dtype
+        )
+        neighbor_valid = torch.ones(
```
Comment (xtk8532704) [inline planner_metrics/object_avoidance.py]: **[P1] Padded neighbor timesteps are treated as valid actors.** `neighbor_agents_future` uses zero coordinates for invalid timesteps, but this all-ones mask passes those padded entries into the OBB collision geometry as degenerate actors at the origin. This can corrupt minimum clearance or produce false collision failures. Please build the per-timestep validity mask from nonzero future coordinates, consistently with `planner_metrics/aggregate.py`.

File: diffusion_planner/diffusion_planner/scenario_based_open_loop/open_loop.py [hunk 3]
```diff
@@ -191,19 +207,20 @@ def run_scenario_based_open_loop_validation(
                             else value
                             for key, value in raw_inputs.items()
                         }
+                        # Named after the source NPZ (rather than a bare sample
+                        # index) so the PNG is identifiable without cross-
+                        # referencing details.jsonl; the title matches.
+                        npz_stem = Path(source_npz).stem
+                        png_path = visualization_root / metric_name / f"{npz_stem}.png"
```
Comment (xtk8532704) [inline diffusion_planner/diffusion_planner/scenario_based_open_loop/open_loop.py]: **[P2] The visualization filename is not unique.** Inputs such as `site_a/scene.npz` and `site_b/scene.npz` produce the same output path, so the later image overwrites the earlier one and both JSONL records point to the same PNG. Please include `sample_index` or a path hash in the filename.

**Review by xtk8532704 — CHANGES_REQUESTED** [review]
I found three issues that should be addressed before merging. I also ran the added scenario/open-loop unit tests locally: 17 tests passed, but the current tests do not cover the integration, padded-timestep, or duplicate-filename cases described in the inline comments.

-- generated by codex

File: diffusion_planner/diffusion_planner/scenario_based_open_loop/open_loop.py (hunk 1 above)
Comment (TatsukiNishimura) [reply diffusion_planner/diffusion_planner/scenario_based_open_loop/open_loop.py:217]: deleted. no need to use sample count.

**Review by xtk8532704 — APPROVED** [review]
Reviewed the three requested fixes: training-time sample_count KeyError, padded neighbor timestep masking, and visualization filename collisions. All are addressed in the latest commits; targeted tests pass.
